src/source/src/manager.rs

```diff
@@ -130,27 +132,53 @@ pub struct MemSourceManager {
 #[async_trait]
 impl SourceManager for MemSourceManager {
     fn get_source(&self, table_id: &TableId) -> Result<SourceDesc> {
-        let sources = self.get_sources()?;
-        sources.get(table_id).cloned().ok_or_else(|| {
+        let inner = self.inner.lock();
+        inner.sources.get(table_id).cloned().ok_or_else(|| {
             InternalError(format!("Get source table id not exists: {:?}", table_id)).into()
         })
     }
 
-    fn drop_source(&self, table_id: &TableId) -> Result<()> {
-        let mut sources = self.get_sources()?;
-        ensure!(
-            sources.contains_key(table_id),
-            "Source does not exist: {:?}",
-            table_id
-        );
-        sources.remove(table_id);
-        Ok(())
+    fn insert_table_source(
+        &self,
+        table_id: &TableId,
+        info: &TableSourceInfo,
+    ) -> Result<SourceDesc> {
+        let mut inner = self.inner.lock();
+        let actor_num = inner.source_actor_num.entry(*table_id).or_insert(0usize);
+        *actor_num += 1;
+        let desc = inner.sources.entry(*table_id).or_insert_with(|| {
+            let columns = info
+                .columns
+                .iter()
+                .cloned()
+                .map(|c| ColumnDesc::from(c.column_desc.unwrap()))
+                .collect_vec();
+            let row_id_index = info.row_id_index.as_ref().map(|index| index.index as _);
+            let pk_column_ids = info.pk_column_ids.clone();
+
+            // Table sources do not need columns and format
+            SourceDesc {
+                columns: columns.iter().map(SourceColumnDesc::from).collect(),
+                source: Arc::new(SourceImpl::Table(TableSource::new(columns))),
+                format: SourceFormat::Invalid,
+                row_id_index,
+                pk_column_ids,
+                metrics: self.metrics.clone(),
+            }
+        });
+        Ok(desc.clone())
     }
 
-    fn clear_sources(&self) -> Result<()> {
-        let mut sources = self.get_sources()?;
-        sources.clear();
-        Ok(())
+    fn try_drop_source(&self, source_id: &TableId) {
+        let mut inner = self.inner.lock();
+        let actor_num = inner
+            .source_actor_num
+            .get_mut(source_id)
+            .expect("double release in local source manager!");
```

Comment:
Is there any edge case to be considered for recovery? cc @yezizp2012

Reply:
I guess no, except that we should clean all source infos when drop all actors.